### PR TITLE
Add test for Delta XDS vs SotW

### PR DESCRIPTION
**Pull request by howardjohn**
This adds a new test -- disabled by default -- which adds a bunch of consistency checks in Delta XDS by comparing responses to SotW responses. This is intended to have 0 impact on production code, as all changes are behind a `UNSAFE` flag.

These tests have found a number of bugs in the past, as I have maintained this in my fork for a while. I think its useful to upstream though.

### CODE REVIEW

File: pilot/pkg/xds/eds.go [hunk 1]
```diff
@@ -551,7 +553,7 @@ func (eds *EdsGenerator) buildEndpoints(proxy *model.Proxy,
 	req *model.PushRequest,
 	w *model.WatchedResource) (model.Resources, model.XdsLogDetails) {
 	var edsUpdatedServices map[string]struct{}
-	if !req.Full {
+	if !req.Full || onlyEndpointsChanged(req) {
```
Comment (howardjohn) [inline pilot/pkg/xds/eds.go:556]: NOTE: I built this on top of https://github.com/istio/istio/pull/37974 to make things simpler. I will back it out if we don't merge that, and put a hold on this PR unti lthat one lands

File: pilot/pkg/model/context.go [hunk 2]
```diff
@@ -342,6 +342,10 @@ type WatchedResource struct {
 
 	// LastSent tracks the time of the generated push, to determine the time it takes the client to ack.
 	LastSent time.Time
+
+	// LastMessage tracks the contents of the last push.
+	// This field is extremely expensive to maintain and is typically disabled
+	LastMessage Resources
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/context.go:348]: LastResources

File: pilot/pkg/xds/delta.go [hunk 3]
```diff
@@ -560,3 +594,126 @@ func extractNames(res []*discovery.Resource) []string {
...
+			log.Errorf("%s: TEST for node:%s missed deletions: %v %v", v3.GetShortType(w.TypeUrl), con.proxy.ID, details, missedDeletes)
+		}
+		if len(missedChanges) > 0 {
+			log.Errorf("%s: TEST for node:%s missed changes: %v %v", v3.GetShortType(w.TypeUrl), con.proxy.ID, details, missedChanges)
+		}
+		if len(extraChanges) > 0 && usedDelta { // TODO flag to log even when !usedDelta, since it means we didn't use delta where we could have
+			log.Infof("%s: TEST for node:%s missed possible optimization: %v. deleted:%v changed:%v", v3.GetShortType(w.TypeUrl), con.proxy.ID, extraChanges, len(gotDeleted), len(gotChanged))
+		}
+	}
+}
```
Comment (hzxuzhonghu) [inline pilot/pkg/xds/delta.go:719]: LGTM if move all these debug function to a separate file

**Review by adiprerepa — APPROVED** [review]
should we dump the delta results (bugs & potential optimizations) to a file of some sort? Could be an Istio report kind of thing, just thinking out loud here.

**Pull request merged**